File: embedded-continuous-query/src/main/java/org/littlewings/infinispan/continuousquery/MyContinuousQueryListener.java
package org.littlewings.infinispan.continuousquery;

import java.util.LinkedHashMap;
import java.util.Map;

import org.infinispan.query.continuous.ContinuousQueryResultListener;

public class MyContinuousQueryListener<K, V> implements ContinuousQueryResultListener<K, V> {
    private Map<K, V> joined = new LinkedHashMap<>();
    private Map<K, Integer> joinCalled = new LinkedHashMap<>();
    private Map<K, Integer> leaveCalled = new LinkedHashMap<>();

    @Override
    public void resultJoining(K key, V value) {
        synchronized (joined) {
            joined.put(key, value);
            incrementNumberOfCalls(joinCalled, key);
        }
    }

    @Override
    public void resultLeaving(K key) {
        incrementNumberOfCalls(leaveCalled, key);
    }

    void incrementNumberOfCalls(Map<K, Integer> countingMap, K key) {
        synchronized (countingMap) {
            countingMap.compute(key, (k, v) -> v == null ? 1 : v + 1);
        }
    }

    public Map<K, V> getJoined() {
        return joined;
    }

    public Map<K, Integer> getJoinCalled() {
        return joinCalled;
    }

    public Map<K, Integer> getLeaveCalled() {
        return leaveCalled;
    }
}

File: embedded-continuous-query-update-support/src/main/scala/org/littlewings/infinispan/continuousquery/MyContinuousQueryListener.scala
package org.littlewings.infinispan.continuousquery

import org.infinispan.query.api.continuous.ContinuousQueryListener

class MyContinuousQueryListener extends ContinuousQueryListener[String, Book] {
  protected[continuousquery] var joined: Int = 0
  protected[continuousquery] var updated: Int = 0
  protected[continuousquery] var leaved: Int = 0

  override def resultJoining(key: String, value: Book): Unit = {
    synchronized(joined += 1)
    println(s"joined, key = ${key}, value = ${value.title}:${value.price}")
  }

  override def resultUpdated(key: String, value: Book): Unit = {
    synchronized(updated += 1)
    println(s"update, key = ${key}, value = ${value.title}:${value.price}")
  }

  override def resultLeaving(key: String): Unit = {
    synchronized(leaved += 1)
    println(s"leaved, key = ${key}")
  }

  def allCount: (Int, Int, Int) = (joined, updated, leaved)
}
